File: lib/rayz/triangle.rb
require_relative "shape"
require_relative "intersection"
require_relative "vector"
require_relative "util"

module Rayz
  class Triangle < Shape
    attr_reader :p1, :p2, :p3, :e1, :e2, :normal

    def initialize(p1:, p2:, p3:)
      super()
      @p1 = p1
      @p2 = p2
      @p3 = p3

      # Precompute edge vectors (convert to Vector explicitly)
      diff1 = p2 - p1
      diff2 = p3 - p1
      @e1 = Vector.new(x: diff1.x, y: diff1.y, z: diff1.z)
      @e2 = Vector.new(x: diff2.x, y: diff2.y, z: diff2.z)

      # Precompute normal (same everywhere on triangle)
      @normal = @e2.cross(@e1).normalize
    end

    def local_intersect(local_ray)
      # Möller-Trumbore algorithm for ray-triangle intersection
      dir_cross_e2 = local_ray.direction.cross(@e2)
      det = @e1.dot(dir_cross_e2)

      # If determinant is near zero, ray lies in plane of triangle
      return [] if det.abs < Util::EPSILON

      f = 1.0 / det
      p1_to_origin_tuple = local_ray.origin - @p1
      p1_to_origin = Vector.new(x: p1_to_origin_tuple.x, y: p1_to_origin_tuple.y, z: p1_to_origin_tuple.z)
      u = f * p1_to_origin.dot(dir_cross_e2)

      # Test u parameter (first barycentric coordinate)
      return [] if u < 0 || u > 1

      origin_cross_e1 = p1_to_origin.cross(@e1)
      v = f * local_ray.direction.dot(origin_cross_e1)

      # Test v parameter (second barycentric coordinate)
      return [] if v < 0 || (u + v) > 1

      # Compute t to find intersection point
      t = f * @e2.dot(origin_cross_e1)

      [Intersection.new(t: t, object: self)]
    end

    def local_normal_at(local_point)
      # Triangle normal is constant across entire surface
      @normal
    end
  end
end

File: python/rayz/triangle.py
from __future__ import annotations

from rayz.constants import EPSILON
from rayz.intersection import Intersection
from rayz.shape import Shape
from rayz.tuple import Vector


class Triangle(Shape):
    def __init__(self, p1, p2, p3) -> None:
        super().__init__()
        self.p1 = p1
        self.p2 = p2
        self.p3 = p3
        d1 = p2 - p1
        d2 = p3 - p1
        self.e1 = Vector(d1.x, d1.y, d1.z)
        self.e2 = Vector(d2.x, d2.y, d2.z)
        self.normal = self.e2.cross(self.e1).normalize()

    def local_intersect(self, ray) -> list:
        dir = Vector(ray.direction.x, ray.direction.y, ray.direction.z)
        dir_cross_e2 = dir.cross(self.e2)
        det = self.e1.dot(dir_cross_e2)
        if abs(det) < EPSILON:
            return []
        f = 1.0 / det
        p1_to_origin = ray.origin - self.p1
        p1o = Vector(p1_to_origin.x, p1_to_origin.y, p1_to_origin.z)
        u = f * p1o.dot(dir_cross_e2)
        if u < 0 or u > 1:
            return []
        origin_cross_e1 = p1o.cross(self.e1)
        v = f * dir.dot(origin_cross_e1)
        if v < 0 or (u + v) > 1:
            return []
        t = f * self.e2.dot(origin_cross_e1)
        return self._make_intersections(t, u, v)

    def _make_intersections(self, t, u, v) -> list:
        return [Intersection(t, self)]

    def local_normal_at(self, point, hit=None) -> Vector:
        return self.normal

    def bounds(self):
        from rayz.bounds import Bounds
        from rayz.tuple import Point
        min_x = min(self.p1.x, self.p2.x, self.p3.x)
        min_y = min(self.p1.y, self.p2.y, self.p3.y)
        min_z = min(self.p1.z, self.p2.z, self.p3.z)
        max_x = max(self.p1.x, self.p2.x, self.p3.x)
        max_y = max(self.p1.y, self.p2.y, self.p3.y)
        max_z = max(self.p1.z, self.p2.z, self.p3.z)
        return Bounds(Point(min_x, min_y, min_z), Point(max_x, max_y, max_z))

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Triangle):
            return NotImplemented
        return self.p1 == other.p1 and self.p2 == other.p2 and self.p3 == other.p3
